content/articles/web-apps.md

```diff
@@ -0,0 +1,108 @@
+- User does not need direct access to large datasets required for code execution
+- Can give researchers quick means to assess utility of a model before putting in effort to download, configure and run locally.
+
+
+## What about Shiny?
+Shiny is a popular choice for researchers wishing to provide web access to their models developed in R, but it has some drawbacks.
+
+Pros:
+- rapid web application development within R
+- easy to get started
+- powerful presentation features
+
+Cons:
+- difficult/impossible to automate tests
+- slow
+- does not scale easily
+- limited capabilities for many application features e.g. user model
+
+
+Shiny's limitations mean that it may not be a good choice for a final application to present to users, but could be a useful prototyping tool, allowing the research team to develop ideas for how an application's user interface should be structured and what options it should provide. The RESIDE team could then migrate the prototype Shiny app's functionality to a full web application.
+
+However, in cases where a simple application is all that is required, with no complex interactivity or scalability requirements, a Shiny application may be the best option for the final application too.
+
+# Components of a full research web application
+
+Our web application installations typically consist of three main components:
+- research code - provided by the research team
+- an interface layer - this provides a way for the web application to provide inputs and receive outputs from the research code. It is usually written in the same language as the research code and invokes it directly, or via a task queue. The interface layer usually provides a local HTTP API which is accessible to the web application
+- web application, consisting of two parts:
+    - back end: e.g kotlin or express.js, which forwards requests to the interface layer, and can do various bits of housekeeping, like managing persistent user data or uploaded files.
+    - front end: e.g Vue.js or React which is loaded in the user's browser and provides the user interface and visualisations.
+
+We run all of these components within docker containers to provide reliable portability between installation environments. Typically the front and back ends of the web applications will go into one container, and the research code and interface layer will go into another. We'll run a web proxy in a third container, and there may be additional containers for databases, task queues etc.
+
+
+# Pre-requisites
+
+In order to support this architecture, we require the research code to have the following features so that we can build a stable interface to it:
+
+- Source-controlled in github: we maintain the source code for our applications in github repositories, and we require research code to be available in github too, so that we can develop against it, apply automated tests to it within continuous integration and deploy it using docker.
```

Comment:
```suggestion
- Source-controlled in git: we maintain the source code for our applications in git repositories, and we require research code to be available in github too, so that we can develop against it, apply automated tests to it within continuous integration and deploy it using docker.
```